### PR TITLE
Use static OpenSSL linking

### LINKED ISSUE #4: Consider statically linking OpenSSL into erts
We have the ability to statically link libs to erts, such as openssl. While this may have some neat security properties, it's more about less fragility in runners. That is to say if we install, build, and link in libraries users won't have to worry about something breaking underneath them. 

Currently, and as part of the initial work, we're depending on brew to install openssl. This also means, that users will have to use brew to install openssl. This can be fragile. 

At the very least we could build openssl ourselves and bundle it, but static linking sounds like the way to go IMHO.

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
# Description

Closes #4.

## Further considerations

ℹ️ I wouldn't mind downloading and compiling OpenSSL, but wouldn't that mean we'd have to have a matrix for "OpenSSL vsn to darwin vsn"? Or should we just go with 1.1 for now and care about that later?

⚠️ I'm gonna go over https://github.com/erlang/otp/issues/4389 with open 👀 and (potentially) update our code as per [Sverker's suggestion](https://github.com/erlang/otp/issues/4389#issuecomment-776587286). @starbelly, do you know if `--disable-dynamic-ssl-lib` is honoured these days, though, and configuration doesn't fall back to dynamic linking if static fails?

## Findings as per OTP release notes

- 24.0 added /usr/local/opt/openssl to the openssl configure search path
  - this helps us some
- 24.0 has configure scripts in crypto and erts ... fail if a requested feature cannot be enabled... It is now better at finding usable OpenSSL libraries
  - we need to find out if this also means that requesting dynamic ssl lib to be disabled also fails in case static linking is not possible (at a point in time it'd fall back to dynamic linking, even if explicitly disabled)
- 24.2 states the crypto app in OTP can now be compiled, linked and used with the new OpenSSL 3.0 cryptolib, (though) this release (is) *not recommended* for other usages than experiments and alpha testing
  - no requirement for change here, for us, just history
- 25.1 states Crypto is now considered to be usable with the OpenSSL 3.0 cryptolib for production code
  - this means that if we're to have a openssl vs. otp matrix we can start linking to 3.0 from 25.1 on

## Further considerations

To move this pull request forward we need to think about:

1. adding a test like https://github.com/jelly-beam/otp-macos/pull/89#issuecomment-1742231421
2. having an issue to notify us in case of failed builds
3. https://github.com/jelly-beam/otp-macos/pull/89#issuecomment-1744200210
4. moving our lowest target to 25.1 (to go for OpenSSL 3.x)

Once the pull request is merged (or prior to that):

1. remove all tags
2. remove all releases
3. remove all `[automation]` commits from the main branch

----

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/jelly-beam/otp-macos/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: .github/workflows/release.sh [hunk 1]
```diff
@@ -102,6 +109,11 @@ kerl_configure() {
     KERL_BUILD_DOCS=yes
     export KERL_BUILD_DOCS
 
+    export_kerl_configuration_option "--disable-dynamic-ssl-lib"
+    local with_ssl
+    with_ssl="$(command -v openssl)/../.."
+    export_kerl_configuration_option "--with-ssl=${with_ssl}"
```
Comment (paulo-ferraz-oliveira) [inline .github/workflows/release.sh:115]: I wanna start simple (and it's not often I find other libs being statically linked, e.g. like `zlib` - probably they've been less prone to issues, over the years) and then maybe we can build bigger, but once we do, yeah, we have to start thinking about versioning, which is a "brave new world" in itself.

**Comment by paulo-ferraz-oliveira** [comment]
Once this is approved+merge we'll probably have to remove all:
* tags
* release
* [automation] commits in `main`

**Comment by paulo-ferraz-oliveira** [comment]
Now failing because of [this](https://github.com/jelly-beam/otp-macos/actions/runs/6367117393/job/17285440230?pr=89#step:3:1233). `kerl` is hard-coding openssl to 1.1 when it's otherwise not specific, and the runners all have 1.1 by default, so I'm missing some context here...

**Comment by paulo-ferraz-oliveira** [comment]
f383ae6 does the same as Kerl, only explicitly. I can't find information on whether `with-ssl` actually disables OpenSSL dynamic linking or not, but that's the only difference we have now to what Kerl does by default. Thoughts?

**Comment by starbelly** [comment]
I just tried static linking with openssl v3 and it worked fine. Albeit with v3 and OTP main (27). Do you want to add a test for this?

```
erl -eval '<<184,87,73,195,111,191,113,10,223,20,12,127,172,13,194,218,18,230,247,8,9,206,243,225,62,72,189,77,63,236,55,72>> = crypto:hash(sha256, <<"Eh?">>), io:format("Basic crypto works!", []), erlang:halt().' -noshell
```

**Comment by starbelly** [comment]
> do you know if `--disable-dynamic-ssl-lib` is honoured these days, though, and configuration doesn't fall back to dynamic linking if static fails?

Will have to test that. I have tested what you have locally with OTP 26 and openssl 3.x and that works fine. In addition to having a test to ensure crypto is functioning, we probably also want to use `otool` to check that you do not see a link to openssl (e.g, `otool -L lib/crypto-5.3/priv/lib/crypto.so`). Easier would be to check at the build site (e.g., `otool -L lib/crypto/priv/lib/aarch64-apple-darwin23.0.0/crypto.so`).

**Comment by paulo-ferraz-oliveira** [comment]
Perfect, so static linking is in place, with this pull request 👍

**Comment by paulo-ferraz-oliveira** [comment]
Rebasing onto the main branch, and force-pushing, to resolve merge conflicts.

**Comment by paulo-ferraz-oliveira** [comment]
We should discuss this again soon, @starbelly. I think we can/should release a version even if it's not perfect, and then let the FOSS world decide. If it's deemed insecure we can either act on specific proposed solution or accept pull requests if they happen. Cheers.

**Comment by starbelly** [comment]
> We should discuss this again soon, @starbelly. I think we can/should release a version even if it's not perfect, and then let the FOSS world decide. If it's deemed insecure we can either act on specific proposed solution or accept pull requests if they happen. Cheers.

I agree, at this point I have less of a concern. What do we need to do to move forward/

**Comment by paulo-ferraz-oliveira** [comment]
> What do we need to do to move forward/

Glad you ask. Lemme fetch my notes...

* we seem to agree on "going simple, for now"
* we seem to agree that "static linking is Ok, for now"
* ⚠️ I need a way to test (in CI, thus programmatically) if the linking was indeed static: ideas?
* we have a couple of identified issues, by I can deal with those after this one's taken care of
(really glad you had time to get back to this, even if for a short while, @starbelly)

**Comment by starbelly** [comment]
> * I need a way to test (in CI, thus programmatically) if the linking was indeed static: ideas?

I suppose, the main test would be to ensure crypto works, which could be done with an eval or run a test project in it with some code that calls out to crypto. Though, as you pointed out, we also need to ensure that it is actually statically linked (compilation can play out, but ssl didn't statically link in the end), I'll have to recheck, but probably have to use `otool` and/or that some files exist. I need to go back and look do a static link run on my local machine to see the results, been a while. 

I do very much like the idea of going full on experimental with it, as suggested. There will be bugs :)

**Comment by starbelly** [comment]
Refreshed myself a bit, so out of the box static linking works like a charm if it's openssl < 3.0. At least with openssl 3.0 as installed via brew, does not work, the message `Static linking with OpenSSL 3.0 *MAY* require special configuring of the cryptolib`.  I suspect per that message it's an issue with openssl as installed via brew, so if we want to support openssl 3, then we probably simply rely on homebrew. WIll have to compile and install openssl 3 ...

**Comment by paulo-ferraz-oliveira** [comment]
Other possibilities:

* we use 1.1 for everything, for now, and think about 3.0 if it surfaces
* we only start generating images from 25.1, use 3.0 for now, and think about 3.0+ if it surfaces

**Comment by paulo-ferraz-oliveira** [comment]
I'm bumping the current code (in this pull request) to use 3.0 and check the result.

**Review by starbelly — APPROVED** [review]

**Pull request merged**